python/cudf/setup.py

```diff
@@ -12,6 +14,74 @@
 
 import versioneer
 
+
+def cuda_detect(cuda_home):
+    """Attempt to detect the version of CUDA present in the operating system.
+    On Windows and Linux, the CUDA library is installed by the NVIDIA
+    driver package, and is typically found in the standard library path,
+    rather than with the CUDA SDK (which is optional for running CUDA apps).
+    On macOS, the CUDA library is only installed with the CUDA SDK, and
+    might not be in the library path.
+    Returns: version string (Ex: '9.2') or None if CUDA not found.
+    """
+
+    system = sys.platform
+    print("system", system)
+    if system == "darwin":
+        lib_filenames = [
+            "libcuda.dylib",  # check library path first
+            os.path.join(cuda_home, "lib/libcuda.dylib"),
+        ]
+    elif system == "linux":
+        lib_filenames = [
+            os.path.join(cuda_home, "lib64/libcudart.so"),
+        ]
+    elif system == "win32":
+        lib_filenames = ["nvcuda.dll"]
+    else:
+        return None  # CUDA not available for other operating systems
+
+    if system == "wind32":
+        dll = ctypes.windll
+    else:
+        dll = ctypes.cdll
+    libcuda = None
+    print("lib_filenames", lib_filenames)
+    for lib_filename in lib_filenames:
+        try:
+            libcuda = dll.LoadLibrary(lib_filename)
+            break
+        except Exception:
+            pass
+    print("libcuda", libcuda)
+    if libcuda is None:
+        return None
+
+    # Get CUDA version
+    try:
+        cudaRuntimeGetVersion = libcuda.cudaRuntimeGetVersion
+        version_int = ctypes.c_int(0)
+        ret = cudaRuntimeGetVersion(ctypes.byref(version_int))
+        print("ret", ret)
+        if ret == 100:
+            cuda_version = os.environ.get("CUDA_VERSION", None)
+            if cuda_version is None:
+                print("ENV_VARS", os.environ)
+                raise TypeError("Please set CUDA_VERSION environment variable")
+
+            print("cuda_version", cuda_version)
+            return ".".join(cuda_version.split(".")[:2])
+        elif ret != 0:
+            return None
+
+        # Convert version integer to version string
+        value = version_int.value
+        return "%d.%d" % (value // 1000, (value % 1000) // 10)
+    except Exception as e:
+        print("Exception", e.__str__())
+        return None
```

Comment:
This is going to detect the installed driver version, whereas we would want to link with the stub library in `${CUDA_HOME}/lib64/stubs` instead.

Now we can't issue the command to that stub library since it's just a stub, so maybe we're better off just pulling the `#define CUDA_VERSION #####` line from `cuda.h` instead.

Reply:
> This is going to detect the installed driver version, whereas we would want to link with the stub library in `${CUDA_HOME}/lib64/stubs` instead.

This(line 62 to 64) is currently detecting cuda runtime version, if that fails because of no GPU on machine it is fetching from `CUDA_VERSION` environment variable. 


> Now we can't issue the command to that stub library since it's just a stub, so maybe we're better off just pulling the `#define CUDA_VERSION #####` line from `cuda.h` instead.

So do you recommend this as a fallback if runtime api fails to return version or just use this as the only approach to get the CUDA_VERSION? There is also one more way from `CUDA_HOME/version.txt` which is a simple text file with CUDA version in a more easily parsable format like:

```
pgali@dt07:/nvme/0/pgali/cudf$ cat /usr/local/cuda/version.txt 
CUDA Version 10.2.89
```

Reply:
I was thinking about something like a conda package in the future which would ship the libraries / headers in a conda package and wouldn't naturally have a nice place for this `version.txt` file to live persay where the `cuda.h` option would guarantee it works for any future packaging mechanism of CUDA.